From 222929ef7c88c099901bcd569c3da6693bedd26b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Markus=20H=C3=A4rer?= <markus.haerer@gmx.net>
Date: Sun, 11 Oct 2026 00:43:52 +0200
Subject: [PATCH] [skin] Notify windows when skin timers change

Skin conditions can depend on skin timers, e.g. Estuary's top bar
overlay hides after Skin.TimerElapsedSecs(1109_topbaroverlay) reaches
5 seconds. A timer changes without any other GUI event, so with smart
redraw the windows using it were not processed and the overlay stayed
visible while paused.

Send GUI_MSG_STATE_CHANGED when a timer's running state or elapsed
whole seconds change, so the windows are processed again.

(cherry picked from commit b0deb27f586d8be82198ee3653181242e78415be)
---
 xbmc/addons/gui/skin/SkinTimer.cpp        | 10 ++++++++++
 xbmc/addons/gui/skin/SkinTimer.h          | 10 ++++++++++
 xbmc/addons/gui/skin/SkinTimerManager.cpp | 11 +++++++++++
 3 files changed, 31 insertions(+)

diff --git a/xbmc/addons/gui/skin/SkinTimer.cpp b/xbmc/addons/gui/skin/SkinTimer.cpp
index d91c0f5..e8324eb 100644
--- a/xbmc/addons/gui/skin/SkinTimer.cpp
+++ b/xbmc/addons/gui/skin/SkinTimer.cpp
@@ -65,6 +65,16 @@ bool CSkinTimer::VerifyStopCondition() const
   return m_stopCondition && m_stopCondition->Get(INFO::DEFAULT_CONTEXT);
 }
 
+bool CSkinTimer::CheckStateChanged()
+{
+  const bool running = IsRunning();
+  const int elapsedSeconds = static_cast<int>(GetElapsedSeconds());
+  const bool changed = running != m_lastRunning || elapsedSeconds != m_lastElapsedSeconds;
+  m_lastRunning = running;
+  m_lastElapsedSeconds = elapsedSeconds;
+  return changed;
+}
+
 const std::string& CSkinTimer::GetName() const
 {
   return m_name;
diff --git a/xbmc/addons/gui/skin/SkinTimer.h b/xbmc/addons/gui/skin/SkinTimer.h
index 47f6eb8..ed2fa52 100644
--- a/xbmc/addons/gui/skin/SkinTimer.h
+++ b/xbmc/addons/gui/skin/SkinTimer.h
@@ -104,6 +104,12 @@ public:
   */
   bool VerifyStopCondition() const;
 
+  /*! \brief Checks if the timer state that skins can see (running state and elapsed whole
+  * seconds) changed since the last call
+  * \return true if the state changed, false otherwise
+  */
+  bool CheckStateChanged();
+
 private:
   /*! \brief Called when this timer is started */
   void OnStart() const;
@@ -125,4 +131,8 @@ private:
   CGUIAction m_stopActions;
   /*! if the timer should be reset on start (or just resumed) */
   bool m_resetOnStart{false};
+  /*! The running state at the last CheckStateChanged() call */
+  bool m_lastRunning{false};
+  /*! The elapsed whole seconds at the last CheckStateChanged() call */
+  int m_lastElapsedSeconds{0};
 };
diff --git a/xbmc/addons/gui/skin/SkinTimerManager.cpp b/xbmc/addons/gui/skin/SkinTimerManager.cpp
index eb41b18..781b3f7 100644
--- a/xbmc/addons/gui/skin/SkinTimerManager.cpp
+++ b/xbmc/addons/gui/skin/SkinTimerManager.cpp
@@ -9,8 +9,11 @@
 #include "SkinTimerManager.h"
 
 #include "GUIInfoManager.h"
+#include "ServiceBroker.h"
 #include "guilib/GUIAction.h"
 #include "guilib/GUIComponent.h"
+#include "guilib/GUIMessage.h"
+#include "guilib/GUIWindowManager.h"
 #include "utils/StringUtils.h"
 #include "utils/XBMCTinyXML2.h"
 #include "utils/log.h"
@@ -217,6 +220,7 @@ void CSkinTimerManager::Stop()
 
 void CSkinTimerManager::Process() const
 {
+  bool changed = false;
   for (const auto& [key, val] : m_timers)
   {
     const std::unique_ptr<CSkinTimer>::pointer timer = val.get();
@@ -232,5 +236,12 @@ void CSkinTimerManager::Process() const
     {
       timer->Reset();
     }
+    changed |= timer->CheckStateChanged();
   }
+
+  // skin conditions on timers change without any other GUI event, so windows
+  // that are only processed when dirty need to be notified
+  if (changed)
+    CServiceBroker::GetGUI()->GetWindowManager().SendMessage(GUI_MSG_NOTIFY_ALL, 0, 0,
+                                                             GUI_MSG_STATE_CHANGED);
 }
-- 
2.53.0

